Store replaced element's aspect ratio into its own struct.

sammygill opened the pull request:
#### 90b0d80278b29c29d422dc253e081ad689960056
<pre>
Store replaced element&apos;s aspect ratio into its own struct.
<a href="https://bugs.webkit.org/show_bug.cgi?id=248096">https://bugs.webkit.org/show_bug.cgi?id=248096</a>
rdar://102525236

Reviewed by Alan Baradlay.

When replaced element&apos;s compute their aspect ratio, the value is stored
inside of a double This value is computed fairly deep in the
computeIntrinsicRatioInformation call stack and then is propagated back
up to where it will be used. This has been causing some slight
issues with precision when making other changes to replaced element
sizing code that uses the aspect ratio.

To help with this, this patch changes how the aspect ratio is stored by
instead storing each dimension that corresponds with the aspect ratio
into a FloatSize. This variable is then propagated back up the same call
stack the same way and then the caller can determine how they want to
use the aspect ratio (compute into a double or use the individual
dimensions).

For this patch there is no functional change because we will store the
value in a double regardless. However, we do seem to progress on a test
(fast/replaced/width100percent-image). The new sizes match how other
browsers are sizing the image. This is likely due to the change in
RenderReplaced::computeIntrinsicRatioInformation. The old version was
computing the intrinsicRatio by dividing two floats. The new version
is storing the dimensions in FloatSize, but the aspect ratio is being
computed by casting one of the values in the operation to a double.
aspectRatioDouble was added into FloatSize to help with this.

This patch also introduces aspectRatioLogicalWidth and
aspectRatioLogicalHeight inside RenderStyle. These methods will return
the correct width/height depending on the writing mode.

* LayoutTests/platform/gtk/fast/replaced/width100percent-image-expected.txt:
* LayoutTests/platform/ios/fast/replaced/width100percent-image-expected.txt:
* LayoutTests/platform/mac/fast/replaced/width100percent-image-expected.txt:
* Source/WebCore/platform/graphics/FloatSize.h:
(WebCore::FloatSize::aspectRatioDouble const):
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::computeIntrinsicRatioInformation const):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::computeIntrinsicRatioInformation const):
* Source/WebCore/rendering/RenderImage.h:
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::computeAspectRatioInformationForRenderBox const):
(WebCore::RenderReplaced::computeIntrinsicAspectRatio const):
(WebCore::RenderReplaced::computeIntrinsicRatioInformation const):
(WebCore::RenderReplaced::computeReplacedLogicalWidth const):
(WebCore::resolveHeightForRatio):
(WebCore::RenderReplaced::computeReplacedLogicalHeight const):
* Source/WebCore/rendering/RenderReplaced.h:
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::aspectRatioLogicalWidth const):
(WebCore::RenderStyle::aspectRatioLogicalHeight const):
* Source/WebCore/rendering/svg/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::computeIntrinsicRatioInformation const):
* Source/WebCore/rendering/svg/LegacyRenderSVGRoot.h:
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::computeIntrinsicRatioInformation const):
* Source/WebCore/rendering/svg/RenderSVGRoot.h:

Canonical link: <a href="https://commits.webkit.org/257430@main">https://commits.webkit.org/257430@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4bc68d5062e8058389b48bd6b19b97faaaa3311a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98870 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8081 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32098 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108285 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168542 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102825 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8635 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85448 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91404 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106271 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104553 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6544 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90170 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33575 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88366 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21515 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76431 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1992 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23031 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1901 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45438 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5109 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6857 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42505 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3300 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -610,7 +610,15 @@ override;
-    virtual void computeIntrinsicRatioInformation(FloatSize& /* intrinsicSize */, double& /* intrinsicRatio */) const { }
+
+    struct AspectRatio {
+        double logicalWidth = 0;
```
Should these be LayoutUnits at this point? Which means this struct is really just a LayoutSize?

sammygill replied:
I believe when I first tried that there were certain tests that were starting to fail due to the ` m_value = clampToInteger(value * kFixedPointDenominator);` calculation but I will double check.

alanbaradlay replied:
which is completely fine, they probably just need rebaseline (unless they are WPT regressions, in which case they need a bit of debugging). We generally rely on LayoutUnit's precision (however limited it is) in (non-inline type of) layout.

sammygill replied:
So changing to LayoutSize is mostly fine, the only test that is potentially concerning is `web-platform-tests/html/rendering/replaced-elements/attributes-for-embedded-content-and-images/img-aspect-ratio.html`

The test sets the width attribute to 0.8 and height attribute to 0.2 on an image element. The loss of information here eventually ends up in an aspect ratio of 4.25249169435216.

Storing the doubles (the way it is right now in the PR) helps avoid this issue, but the only concern I have with using them is that the LayoutUnit operators would also return doubles without checking for overflow (there is a WPT test checking this and using the LayoutSizes and the individual dimensions in this case results in a failure). Seems like it may need a little bit of work to get a correct solution

sammygill replied:
After looking into this a bit more and playing around with the code/tests, I'm not sure if LayoutSize is the correct type to use in this context. I ended up using FloatSize, which seems to work fine, since the original values are crucial in the `img-aspect-ratio.html` test. If we convert these values into LayoutUnits we lose this information and I'm not sure we can pass the test without it, but let me know if any of you disagree or have other thoughts.

sammygill committed:
Store replaced element's aspect ratio into its own struct.
https://bugs.webkit.org/show_bug.cgi?id=248096
rdar://102525236

Reviewed by Alan Baradlay.

When replaced element's compute their aspect ratio, the value is stored
inside of a double This value is computed fairly deep in the
computeIntrinsicRatioInformation call stack and then is propagated back
up to where it will be used. This has been causing some slight
issues with precision when making other changes to replaced element
sizing code that uses the aspect ratio.

To help with this, this patch changes how the aspect ratio is stored by
instead storing each dimension that corresponds with the aspect ratio
into a FloatSize. This variable is then propagated back up the same call
stack the same way and then the caller can determine how they want to
use the aspect ratio (compute into a double or use the individual
dimensions).

For this patch there is no functional change because we will store the
value in a double regardless. However, we do seem to progress on a test
(fast/replaced/width100percent-image). The new sizes match how other
browsers are sizing the image. This is likely due to the change in
RenderReplaced::computeIntrinsicRatioInformation. The old version was
computing the intrinsicRatio by dividing two floats. The new version
is storing the dimensions in FloatSize, but the aspect ratio is being
computed by casting one of the values in the operation to a double.
aspectRatioDouble was added into FloatSize to help with this.

This patch also introduces aspectRatioLogicalWidth and
aspectRatioLogicalHeight inside RenderStyle. These methods will return
the correct width/height depending on the writing mode.

* LayoutTests/platform/gtk/fast/replaced/width100percent-image-expected.txt:
* LayoutTests/platform/ios/fast/replaced/width100percent-image-expected.txt:
* LayoutTests/platform/mac/fast/replaced/width100percent-image-expected.txt:
* Source/WebCore/platform/graphics/FloatSize.h:
(WebCore::FloatSize::aspectRatioDouble const):
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::computeIntrinsicRatioInformation const):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::computeIntrinsicRatioInformation const):
* Source/WebCore/rendering/RenderImage.h:
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::computeAspectRatioInformationForRenderBox const):
(WebCore::RenderReplaced::computeIntrinsicAspectRatio const):
(WebCore::RenderReplaced::computeIntrinsicRatioInformation const):
(WebCore::RenderReplaced::computeReplacedLogicalWidth const):
(WebCore::resolveHeightForRatio):
(WebCore::RenderReplaced::computeReplacedLogicalHeight const):
* Source/WebCore/rendering/RenderReplaced.h:
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::aspectRatioLogicalWidth const):
(WebCore::RenderStyle::aspectRatioLogicalHeight const):
* Source/WebCore/rendering/svg/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::computeIntrinsicRatioInformation const):
* Source/WebCore/rendering/svg/LegacyRenderSVGRoot.h:
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::computeIntrinsicRatioInformation const):
* Source/WebCore/rendering/svg/RenderSVGRoot.h:

Canonical link: https://commits.webkit.org/257430@main
